azavea/echo-locator
# Implement user signup

## Overview

This PR allows for users to sign up for an account with their email address and, on form submit, receive an emailed link to create a profile and sign in. Below the surface, when a user clicks the “create account” button: their email is validated, a new `User` is created, the new `User` is added to the `HouseSeeker` group, and an empty related `UserProfile` is created before sending the email. Appropriate frontend changes were made, including:
- Adding a clickable prompt on the “sign in” page to create an account (and from the “sign up” page a prompt to toggle back to “sign in”). It should be noted that these are not two separate pages/components, but just toggles between appropriate text/actions.
- Updated `loginMessage` payloads to walk through the signup process or give context to any signup errors.

### Notes

- One tricky part of this PR was figuring out how to let a new `User` create their initial `UserProfile`, even though the authorization to reach the `EditProfile` component relies on the presence of a `UserProfile`. The way I chose to get around this, while also differentiating `Users` who have attempted to sign up but haven't actually clicked the email link or who haven't create a `UserProfile`, was to ~~create an intermediary `AccountProfile` type: `{unverifiedUserProfile: boolean}`. Responding with this object from the backend and setting it as the profile seemed simpler than creating a field on the `UserProfile` model and allows `profile !== undefined` while also differentiating between a complete `UserProfile` and `isAnonymous`~~ create an empty `UserProfile` on signup. This allows `profile !== undefined` to pass authentication and will populate/view a default `EditProfile` component as if `isAnonymous` without actually being `isAnonymous`.
- I made a couple adjustments to the `AccountProfile` type as well as some of the conditionals in `EditProfile` to better fit when a new `User` _has_ a profile, but doesn't actually have anything in it to validate. This is namely getting rid of the `key` property since this is a leftover of our Amplify/S3 profile process (I couldn't find any active purpose to it) as well as making `voucherNumber` optional. These changes were also made as part of the open PR #497 so I tried to stay as close as possible to those changes to avoid any conflicts.
- Full account creation depends on the ability to edit a `UserProfile`. However, since there’s an open PR #497 for editing `UserProfile` from the frontend I decided not to touch any of the functionality. This PR simply creates an empty `UserProfile` for new `Users` and, once we have the ability to save edits to `UserProfile`, the absence of the `unverifiedUserProfile` property will signify account creation is "complete".
- I left the email body copy the same as login since it seemed broad enough to cover signup as well. Any extra signup context is included as part of the `loginMessage` after submitting email.

## Testing Instructions

 * `scripts/update` 
 * `scripts/manage test users.tests` to confirm tests function
 * `scripts/server`
 * Navigate to `localhost:9966/`
 * Click the "create an account" text to toggle to the signup form
 * Try and submit an invalid email and confirm account creation fails and error response message appears
 * Try and submit a valid email and confirm email link prints in the console
 * Try and submit the same email again and confirm duplicate account creation fails
 * Navigate to the email link and confirm it loads `/profile` page, and all fields are present but empty
 * Confirm you can't navigate beyond `/profile` without setting required fields
 * Create a new `User` through the admin, and don't create a related `UserProfile`. Login with the new `User` email and confirm still authorized to navigate to the `/profile` page.


Resolves #465
